### PR TITLE
Polish layout (resolve #199, resolve #172, resolve #174, resolve #198)

### LINKED ISSUE #172: Polish layout components
We have layout components for even columns. We need to add one for a 2/3 1/3 setup to work with the new dashboard:

image

All layout components should be updated to the latest spacing as well which is `24px` margin (gutter) between desktop columns horizontally and vertically.

screen shot 2017-03-24 at 4 04 33 pm

image

All layout component columns should continue to stack on mobile.

### LINKED ISSUE #174: Move lesson proposal into requested card component
The requested lessons card should be updated with a new "Propose a new topic" link at the top of the list which takes the instructor to the `/new` route (which includes lesson proposal). Then the lesson proposal card on the dashboard screen.

image

Mock source:
https://egghead.invisionapp.com/share/KQB170ZD6#/screens

### LINKED ISSUE #198: Polish Navigation component
Navigation has been updated with some new styles, colors, and icons:
image

It should continue to work on mobile.

### LINKED ISSUE #199: Polish title card component
Title cards have been updated to place the title + [description] inside the title card component with updated text styles and a lighter background color:

image

image

### PR DESCRIPTION
**Pull request by trevordmiller**
# Changes

- Abstract screen width resizer as a render callback component
- Move proposing lessons inside the requesting lessons card
- Add relative sizing options to the layout columns component
- Other misc styles updated to match latest mocks

# Result For User

polish-layout

---

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -16,7 +16,7 @@
     "react-scripts": "^0.8.4"
   },
   "dependencies": {
-    "egghead-ui": "^4.3.1",
+    "egghead-ui": "^4.4.0",
```
Comment (trevordmiller) [inline package.json:19]: Updated egghead-ui with new icon `types` for navigation

File: src/components/DeviceWidth/index.js [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+      })
+    }
+  }
+
+  render() {
+    const {isMobile} = this.state
+    const {children} = this.props
+    return children(isMobile)
+  }
+}
```
Comment (trevordmiller) [inline src/components/DeviceWidth/index.js:36]: Allows an escape hatch for screen width resizing events

File: src/components/LayoutColumns/index.js [hunk 3]
```diff
@@ -1,12 +1,19 @@
...
         key={index}
-        className={`w-100-ns ${index < items.length - 1 ? 'mr4-ns' : ''}`}
+        className={`
+          ${index < items.length - 1 ? 'mr4-l' : ''}
+        `}
+        style={{
+          flex: relativeSizes
+            ? relativeSizes[index]
+            : 1
+        }}
```
Comment (trevordmiller) [inline src/components/LayoutColumns/index.js:16]: Can now have relative sizes applied to each column:
image

Can still have even columns by default:
image

Columns still stack vertically on mobile

File: src/components/Navigation/index.js [hunk 4]
```diff
@@ -38,146 +43,166 @@ export default class Navigation extends Component {
...
-      </aside>
+                  <Text message='navigation.action' />
+                </Button>
+              </Link>
+            </div>
+
+          </aside>
+        )}
+      </DeviceWidth>
     )
```
Comment (trevordmiller) [inline src/components/Navigation/index.js:206]: Updated navigation styles:
image

Not pixel perfect to mocks, but pretty close using tachyons-egghead. The button styles will be updated in: 
https://github.com/eggheadio/egghead-ui/issues/93

File: src/components/RequestedLessonsCard/components/ProposeLesson/index.js [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+          onClick={this.handleClick}
+          className='ttu'
+        >
+          <IconLabel
+            iconType='add'
+            labelText={<Text message='requestedLessons.proposeLesson.action' />}
+            color='blue'
+          />
+        </div>
+  }
```
Comment (trevordmiller) [inline src/components/RequestedLessonsCard/components/ProposeLesson/index.js:33]: This `ProposeLesson` component now by default renders this:
image

Then when tapped converts to the form:
image

Vojta plans to update the form to be a step-by-step process, but that will be done in the future in another PR and still needs discussion.

File: src/components/TitleCard/index.js [hunk 6]
```diff
@@ -2,16 +2,18 @@ import React from 'react'
...
+      <Heading level='5'>
+        {title}
+      </Heading>
+      {description
+        ? <Paragraph type='small'>
+            {description}
+          </Paragraph>
+        : null
+      }
+    </div>
```
Comment (trevordmiller) [inline src/components/TitleCard/index.js:16]: Updates the title card to put the title/description inside the card with latest mock styles:
image

File: src/index.js [hunk 7]
```diff
@@ -66,26 +66,32 @@ const App = () => {
...
                     {
                       text: <Text message='navigation.chat' />,
                       action: chatUrl,
+                      iconType: 'slack',
                     },
                     {
                       text: <Text message='navigation.logout' />,
                       action: logout,
+                      iconType: 'sign-out',
                     },
```
Comment (trevordmiller) [inline src/index.js:95]: Adds icons to navigation items. Vojta wants to customize these in the future.

image

File: src/screens/Dashboard/index.js [hunk 8]
```diff
@@ -1,30 +1,38 @@
...
+      />
+    </div>
+  : <div>
+      <LayoutColumns items={[
+        <GetPublishedCard instructor={instructor} />,
+        <RequestedLessonsCard instructor={instructor} />,
+        <HelpCard publishedLessons={instructor.published_lessons} />,
+      ]} />
+      <InstructorLessonsCard instructor={instructor} />
+    </div>
```
Comment (trevordmiller) [inline src/screens/Dashboard/index.js:38]: Changes dashboard layout depending on if instructor is published or not:

Unpublished:
image

Published:
image

This matches the latest mock layouts; the individual components inside will be updated in separate PRs.

Comment (trevordmiller) [reply src/screens/Dashboard/index.js:38]: The unpublished view will also be updated with the "next step" setup from the latest mocks in a future PR.

File: src/utils/hardCodedSizes/index.js [hunk 9]
```diff
@@ -0,0 +1,3 @@
+export const mobileMediaQuerySize = 640
+
+export const navigationWidth = 200
```
Comment (trevordmiller) [inline src/utils/hardCodedSizes/index.js:3]: Some hard-coded sizes needed as tachyons has too large of jumps between higher relative numbers (4-9).

File: src/components/DeviceWidth/index.js (hunk 2 above)
Comment (tayiorbeii) [reply src/components/DeviceWidth/index.js:36]: This is a great example of "Code I can read and makes sense what it does, but it would be hard for me to write" 👍

File: src/components/LayoutColumns/index.js (hunk 3 above)
Comment (tayiorbeii) [reply src/components/LayoutColumns/index.js:16]: Looks so good!

**Review by tayiorbeii — APPROVED** [review]

File: src/components/DeviceWidth/index.js (hunk 2 above)
Comment (theianjones) [reply src/components/DeviceWidth/index.js:36]: This is what I was struggling with for the header. I think @DrShpongle is taking the bug I logged. ^ might be useful.

**Review by theianjones — APPROVED** [review]
This looks great!

File: src/components/DeviceWidth/index.js (hunk 2 above)
Comment (trevordmiller) [reply src/components/DeviceWidth/index.js:36]: Thanks :) I love Render Callbacks (AKA function as child); makes it really easy to share stateful logic across components. Here is my snippet for them if interested:
https://github.com/trevordmiller/settings/blob/master/syncfiles/UltiSnips/all.snippets#L119-L134

File: src/components/LayoutColumns/index.js (hunk 3 above)
Comment (trevordmiller) [reply src/components/LayoutColumns/index.js:16]: Thanks :) Props to @vojtaholik. More goodness coming soon: 
https://app.zeplin.io/project.html#pid=58caceae349a0d1341e322a5
